src/Handler/CurlFactory.php

```diff
@@ -446,7 +446,9 @@ private function applyHandlerOptions(EasyHandle $easy, array &$conf): void
                 if (isset($options['proxy'][$scheme])) {
                     $host = $easy->request->getUri()->getHost();
                     if (isset($options['proxy']['no']) && Utils::isHostInNoProxy($host, $options['proxy']['no'])) {
-                        unset($conf[\CURLOPT_PROXY]);
+                        // To support older versions of php
+                        $conf[\CURLOPT_PROXY] = '';
+                        $conf[\CURLOPT_NOPROXY] = $host;
                     } else {
                         $conf[\CURLOPT_PROXY] = $options['proxy'][$scheme];
                     }
```

Comment:
Although this covers the case where our request matches the `['proxy']['no']` configuration (and we want it not to use the proxy), it misses an inverse case. :)

Image that in the environment variables there's `NO_PROXY="foo.bar"` / `no_proxy="foo.bar"`, and we want to configure guzzle to override/ignore this (to proxy everything, or at least proxy requests to `foo.bar`).

If we set:
```
'proxy' => [
    'http' => 'http://example:8080',
    'https' => 'http://example:8080',
    # Send everything through the proxy
    'no' => []
]
```

Then the environment's behavior will still get used and requests to `foo.bar` will (unexpectedly) not go through the proxy.

Perhaps something like this in the else branch?

```suggestion
                    } else {
                        $conf[\CURLOPT_PROXY] = $options['proxy'][$scheme];
                        $conf[\CURLOPT_NOPROXY] = "";
                    }
```

(Related comment on the issue: https://github.com/guzzle/guzzle/issues/3163#issuecomment-2065612465)

Reply:
hi, well, i think a little correction is needed here, cause your code would unset no proxy even if there is no no proxy value added, we only wanna unset if we set it and set it to be empty, but you are correct that we might need a third case:
'proxy' => [
    'http' => 'http://example:8080',
    'https' => 'http://example:8080',
]

This is the correct i think:
```
                    if (isset($options['proxy']['no'])) {
                        if (Utils::isHostInNoProxy($host, $options['proxy']['no'])) {
                            // To support older versions of php
                            $conf[\CURLOPT_PROXY] = '';
                            $conf[\CURLOPT_NOPROXY] = $host;
                        } else if (empty($options['proxy']['no'])) {
                            $conf[\CURLOPT_NOPROXY] = '';
                        }
                    } else {
                        $conf[\CURLOPT_PROXY] = $options['proxy'][$scheme];
                    }
```
What do you think @oostapov ?